src/server/lock_manager/deadlock.rs

```diff
@@ -305,32 +405,181 @@ impl DetectTable {
+    /// waiters on this lock. For each detected deadlocks, returns a tuple
+    /// `(txn_id, key_hash, key, wait_chain, resource_group_tag)`. Note that
+    /// the `key_hash` and `key` differs from the one given in parameter. Same
+    /// as the `detect` function's return value, the key is the one on which
+    /// the transaction aborting by the deadlock blocks another transaction
+    /// in the deadlock circle.
+    fn replace_lock_by_key(
+        &mut self,
+        lock_hash: u64,
+        key: Vec<u8>,
+        old_lock_ts: TimeStamp,
+        new_lock_ts: TimeStamp,
+    ) -> Vec<(TimeStamp, u64, Vec<u8>, Vec<WaitForEntry>, Vec<u8>)> {
+        if old_lock_ts == new_lock_ts {
+            // It's theoretically possible that a transaction releases a lock and then
+            // acquires it again, e.g., a pessimistic transaction performs a
+            // statement retry. From a waiter's perspective, the lock-holding transaction
+            // didn't change, so nothing to do.
+            return vec![];
+        }
+        if key.is_empty() {
+            // Cannot find in the key_reverse_map if the key is missing. Ignore the case.
+            return vec![];
+        }
+
+        // Find a list of transactions that is blocked by `old_lock_ts` on `key`.
+        let locks_and_waiters_of_key = match self.key_reverse_map.get_mut(&key) {
+            Some(v) => v,
+            None => return vec![],
+        };
+        let waiting_txn = match locks_and_waiters_of_key.remove(&old_lock_ts) {
+            Some(v) => v,
+            None => return vec![],
+        };
+
+        let mut result = vec![];
+
+        // For each txn, update their entries in `wait_for_map` to change their blockers
+        // from `old_lock_ts` to `new_lock_ts`.
+        for txn_ts in &waiting_txn {
+            // Unwrap as `wait_for_map` and `key_reverse_map` should be synchronized.
+            let blockers = self.wait_for_map.get_mut(txn_ts).unwrap();
+            let old_locks = blockers.get_mut(&old_lock_ts).unwrap();
+            assert!(old_locks.remove(lock_hash, &key).is_some());
+            let resource_group_tag = old_locks.resource_group_tag.clone();
+            if old_locks.is_empty() {
+                blockers.remove(&old_lock_ts);
+            }
+
+            if let Some((deadlock_key_hash, deadlock_key, wait_chain)) = self
+                .detect_without_consistency_check(
+                    *txn_ts,
+                    new_lock_ts,
+                    lock_hash,
+                    &key,
+                    &resource_group_tag,
+                )
+            {
+                result.push((
+                    *txn_ts,
+                    deadlock_key_hash,
+                    deadlock_key,
+                    wait_chain,
+                    resource_group_tag,
+                ));
+            }
+        }
+
+        self.check_key_reverse_index_consistency();
+        result
```

Comment:
We may need to record this kind of task into `TASK_COUNTER_METRICS` before returning.